src/Data/TypeMap/Internal/Dynamic.hs

```diff
@@ -60,6 +60,17 @@ insert t v (TypeMap m) = TypeMap (Map.insert (typeRep t) (coerce v) m)
     coerce :: Item x t -> Any
     coerce = unsafeCoerce
 
+-- | Update an element indexed by type @t@.
+update
+  :: forall t x proxy
+  .  Typeable t => proxy t -> (Item x t -> Maybe (Item x t)) -> TypeMap x -> TypeMap x
+update t f (TypeMap m) = TypeMap (Map.update (coerceA . f . coerceB) (typeRep t) m)
```

Comment:
Could this be `Map.update (coerce f)`, so this would not create an additional closure around `f`?